gematria/datasets/exegesis_benchmark_lib_test.cc

```diff
@@ -0,0 +1,386 @@
+    LLVMInitializeX86Target();
+    LLVMInitializeX86AsmPrinter();
+    LLVMInitializeX86AsmParser();
+    LLVMInitializeX86Disassembler();
+
+    // Exegesis Setup
+    InitializeX86ExegesisTarget();
+
+    pfm::pfmInitialize();
+  }
+
+  Expected<std::string> getErrorMessage(StringRef JSONBlock) {
+    Expected<json::Value> BlockValue = json::parse(JSONBlock);
+    if (!BlockValue) return BlockValue.takeError();
+
+    Expected<BenchmarkCode> BenchCode =
+        Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 1);
+    if (BenchCode)
+      return llvm::make_error<StringError>(errc::interrupted,
+                                           "Failed to get BenchmarkCode");
+
+    std::string ErrorMessage;
+    handleAllErrors(BenchCode.takeError(),
+                    [&ErrorMessage](StringError &BenchCodeError) {
+                      ErrorMessage = BenchCodeError.getMessage();
+                    });
+
+    return ErrorMessage;
+  }
+
+  Expected<double> benchmark(StringRef JSONBlock) {
+    Expected<json::Value> BlockValue = json::parse(JSONBlock);
+    if (!BlockValue) return BlockValue.takeError();
+
+    Expected<BenchmarkCode> BenchCode =
+        Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 0);
+    if (!BenchCode) return BenchCode.takeError();
+
+    return Benchmark->benchmarkBasicBlock(*BenchCode);
+  }
+};
+
+TEST_F(ExegesisBenchmarkTest, TestParseJSONBlock) {
+  StringRef JSONBlock = R"json(
+  {
+    "Hex": "3b31",
+    "LoopRegister": 51,
+    "MemoryDefinitions": [
+      {
+        "Name": "MEM",
+        "Size": 4096,
+        "Value": 1
+      }
+    ],
+    "MemoryMappings": [
+      {
+        "Address": 86016,
+        "Value": "MEM"
+      }
+    ],
+    "RegisterDefinitions": [
+      {
+        "Register": 54,
+        "Value": 86016
+      },
+      {
+        "Register": 60,
+        "Value": 86016
+      }
+    ]
+  }
+  )json";
+
+  Expected<json::Value> BlockValue = json::parse(JSONBlock);
+  ASSERT_TRUE(static_cast<bool>(BlockValue));
+  Expected<BenchmarkCode> BenchCode =
+      Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 1);
+
+  ASSERT_THAT(BenchCode->Key.Instructions, testing::SizeIs(1));
+  EXPECT_EQ(BenchCode->Key.Instructions[0].getOpcode(), X86::CMP32rm);
+
+  EXPECT_EQ(BenchCode->Key.LoopRegister, 51);
+
+  ASSERT_THAT(BenchCode->Key.MemoryValues, testing::SizeIs(1));
+  EXPECT_EQ(BenchCode->Key.MemoryValues["MEM"].SizeBytes, 4096);
+  EXPECT_EQ(BenchCode->Key.MemoryValues["MEM"].Value, 1);
+
+  ASSERT_THAT(BenchCode->Key.RegisterInitialValues, testing::SizeIs(2));
+  EXPECT_EQ(BenchCode->Key.RegisterInitialValues[0].Register, 54);
+  EXPECT_EQ(BenchCode->Key.RegisterInitialValues[0].Value, 86016);
+  EXPECT_EQ(BenchCode->Key.RegisterInitialValues[1].Register, 60);
+  EXPECT_EQ(BenchCode->Key.RegisterInitialValues[1].Value, 86016);
+}
+
+TEST_F(ExegesisBenchmarkTest, TestParseJSONNoHex) {
+  Expected<std::string> ErrorMessage = getErrorMessage(R"json(
```

Comment:
Nit: making this a multiline string seems unnecessary, I think `getErrorMessage("{}")` would be more readable.